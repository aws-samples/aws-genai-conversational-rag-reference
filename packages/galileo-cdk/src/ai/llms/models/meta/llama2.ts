/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { LLAMA2_ADAPTER, LLAMA2_ENDPOINT_KWARGS, LLAMA2_KWARGS } from '@aws/galileo-sdk/lib/models/llms/meta';
import { CfnMapping, Stack } from 'aws-cdk-lib';
import { Construct, IConstruct } from 'constructs';
import { BaseLLMProps } from '../../framework';
import { SageMakerLLM } from '../../framework/sagemaker';

export const LLAMA2_13B_SAGEMAKER_PACKAGE_ARN = {
  'ap-southeast-1':
    'arn:aws:sagemaker:ap-southeast-1:192199979996:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
  'ap-southeast-2':
    'arn:aws:sagemaker:ap-southeast-2:666831318237:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
  'eu-west-1': 'arn:aws:sagemaker:eu-west-1:985815980388:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
  'us-east-1': 'arn:aws:sagemaker:us-east-1:865070037744:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
  'us-east-2': 'arn:aws:sagemaker:us-east-2:057799348421:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
  'us-west-2': 'arn:aws:sagemaker:us-west-2:594846645681:model-package/llama2-13b-v4-c4de6690de6132cb962827bec6ef6811',
} as const;

export function resolveLlama2Base13BPackageArn(scope: IConstruct): string {
  return new CfnMapping(scope, 'Llama2Base13BPackageMapping', {
    lazy: true,
    mapping: Object.fromEntries(Object.entries(LLAMA2_13B_SAGEMAKER_PACKAGE_ARN).map(([key, arn]) => [key, { arn }])),
  }).findInMap(Stack.of(scope).region, 'arn');
}

export interface Llama2Base13BSageMakerProps extends BaseLLMProps {}

export class Llama2Base13BSageMaker extends SageMakerLLM {
  constructor(scope: Construct, id: string, props: Llama2Base13BSageMakerProps) {
    super(scope, id, {
      ...props,
      instanceType: props.instanceType ?? 'ml.g5.12xlarge',
      containerDefinition: {
        modelPackageName: resolveLlama2Base13BPackageArn(scope),
      },
      adapter: LLAMA2_ADAPTER,
      modelKwargs: LLAMA2_KWARGS,
      endpointKwargs: LLAMA2_ENDPOINT_KWARGS,
    });
  }
}
