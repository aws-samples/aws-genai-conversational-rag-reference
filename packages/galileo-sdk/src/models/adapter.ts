/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseSageMakerContentHandler } from 'langchain/llms/sagemaker_endpoint';
import { set, get, isEmpty } from 'lodash';
import { getLogger } from '../common/index.js';
import { BaseChatTemplatePartials, ChatCondenseQuestionPromptRuntime, ChatQuestionAnswerPromptRuntime } from '../prompt/templates/chat/index.js';

const logger = getLogger('models/adapter');

/**
 * @struct
 */
export interface IContentHandlerAdapterInput {
  /**
   * Path in input object to define the prompt value.
   * @see https://lodash.com/docs/#set
   * @default "text_inputs"
   */
  readonly promptKey?: string;
  /**
   * Path in the input object to define the model kwargs.
   * @see https://lodash.com/docs/#set
   * @default undefined - Kwargs will be spread on root
   */
  readonly modelKwargsKey?: string;
  // readonly prepend?: string;
  // readonly append?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapterOutput {
  /**
   * Path of response object to extract as output.
   * @see https://lodash.com/docs/#get
   * @default `[0].generated_text`
   */
  readonly jsonpath?: string;
  /**
   * @default "utf-8"
   */
  readonly encoding?: BufferEncoding;
}
/**
 * @struct
 */
export interface IContentHandlerAdapter {
  readonly contentType?: string;
  readonly accepts?: string;
  readonly input?: IContentHandlerAdapterInput;
  readonly output?: IContentHandlerAdapterOutput;
}

/**
 * @struct
 */
export interface IChatPromptAdapter {
  readonly base?: Partial<BaseChatTemplatePartials>;
  readonly questionAnswer?: ChatQuestionAnswerPromptRuntime;
  readonly condenseQuestion?: ChatCondenseQuestionPromptRuntime;
}
/**
 * @struct
 */
export interface IPromptAdapter {
  readonly chat?: IChatPromptAdapter;
}

export class PromptAdapter implements IPromptAdapter {

  readonly chat?: IChatPromptAdapter;

  constructor(adapter?: IPromptAdapter) {
    this.chat = {
      base: adapter?.chat?.base,
      condenseQuestion: adapter?.chat?.base || adapter?.chat?.condenseQuestion ? {
        ...adapter.chat.condenseQuestion,
        templatePartials: {
          ...adapter.chat.base,
          ...adapter.chat.condenseQuestion?.templatePartials,
        },
      } : undefined,
      questionAnswer: adapter?.chat?.base || adapter?.chat?.questionAnswer ? {
        ...adapter.chat.questionAnswer,
        templatePartials: {
          ...adapter.chat.base,
          ...adapter.chat.questionAnswer?.templatePartials,
        },
      } : undefined,
    };
  }
}

/**
 * @struct
 */
export interface IModelAdapter {
  readonly prompt?: IPromptAdapter;
  readonly contentHandler?: IContentHandlerAdapter;
}

export const DEFAULT_OUTPUT_JSONPATH = '[0].generated_text';

export class AdaptedContentHandler extends BaseSageMakerContentHandler<string, string> {
  readonly contentType: string;
  readonly accepts: string;

  protected readonly promptKey: string;
  protected readonly modelKwargsKey?: string;
  protected readonly inputEncoding: BufferEncoding;

  protected readonly outputJsonpath: string;
  protected readonly outputEncoding: BufferEncoding;

  constructor(adapter?: IContentHandlerAdapter) {
    super();

    this.accepts = adapter?.accepts ?? 'application/json';
    this.contentType = adapter?.contentType ?? 'application/json';

    this.promptKey = adapter?.input?.promptKey ?? 'text_inputs';
    this.modelKwargsKey = adapter?.input?.modelKwargsKey;
    this.inputEncoding = adapter?.input?.encoding ?? 'utf-8';

    this.outputJsonpath = adapter?.output?.jsonpath ?? DEFAULT_OUTPUT_JSONPATH;
    this.outputEncoding = adapter?.output?.encoding ?? 'utf-8';

    if (this.accepts !== 'application/json' || this.contentType != 'application/json') {
      throw new Error(`Only "application/json" is supported at this time: ${this.accepts} / ${this.contentType} not implemented`);
    }
  }

  async transformInput(prompt: string, modelKwargs: Record<string, unknown>): Promise<Uint8Array> {
    const input = {};

    set(input, this.promptKey, prompt);

    if (this.modelKwargsKey) {
      set(input, this.modelKwargsKey, modelKwargs);
    } else {
      // spread kwargs if not key
      Object.assign(input, modelKwargs);
    }

    logger.debug('AdaptedContentHandler#transformInput:', { prompt, modelKwargs, input });

    return Buffer.from(JSON.stringify(input), this.inputEncoding);
  }

  async transformOutput(output: Uint8Array): Promise<string> {
    const response = JSON.parse(Buffer.from(output).toString(this.outputEncoding));
    // extract the path from json
    const result = get(response, this.outputJsonpath);
    logger.debug('AdaptedContentHandler#transformOutput:', { response, result });
    return result;
  }
}

export class ModelAdapter {
  readonly prompt?: IPromptAdapter;

  readonly contentHandler: AdaptedContentHandler;

  readonly isDefault: boolean;

  constructor(config?: IModelAdapter) {
    this.isDefault = config == null || isEmpty(config);

    this.prompt = new PromptAdapter(config?.prompt);

    this.contentHandler = new AdaptedContentHandler(config?.contentHandler);
  }
}
