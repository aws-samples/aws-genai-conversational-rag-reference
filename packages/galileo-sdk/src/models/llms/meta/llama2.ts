/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import type { IModelAdapter } from '../../adapter.js';
import { DEFAULT_MAX_NEW_TOKENS } from '../../constants.js';
import type { Kwargs } from '../../types.js';

export const LLAMA2_ADAPTER: IModelAdapter = {
  prompt: {},
  contentHandler: {
    input: {
      promptKey: 'inputs',
      modelKwargsKey: 'parameters',
    },
    output: {
      jsonpath: '[0].generation',
    },
  },
};

export const LLAMA2_KWARGS: Kwargs = {
  temperature: 0.6,
  top_p: 0.9,
  repetition_penalty: 1.1,
  max_new_tokens: DEFAULT_MAX_NEW_TOKENS,
};

export const LLAMA2_ENDPOINT_KWARGS: Kwargs = {
  CustomAttributes: 'accept_eula=true',
};
