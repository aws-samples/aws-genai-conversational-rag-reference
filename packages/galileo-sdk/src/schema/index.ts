/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
export enum ChainType {
  QA = 'QA',
  CONDENSE_QUESTION = 'CONDENSE_QUESTION',
  CLASSIFY = 'CLASSIFY',
}
