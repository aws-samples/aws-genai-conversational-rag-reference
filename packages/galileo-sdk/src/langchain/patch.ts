/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseMessage } from 'langchain/schema';

// Patch langchain BaseMessage prototype to have "type" getter
// Currently only has _getType() method and we need simple getter for template
Object.defineProperty(BaseMessage.prototype, 'type', {
  get: function () {
    return this._getType();
  },
  configurable: true,
  enumerable: true,
});
