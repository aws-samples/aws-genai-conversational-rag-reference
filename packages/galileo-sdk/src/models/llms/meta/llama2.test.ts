/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
// @ts-ignore - .test files are ignored
import type {} from '@types/jest';
import { AIMessage, HumanMessage, SystemMessage } from 'langchain/schema';
import { LLAMA2_ADAPTER } from './llama2.js';
import { ChatCondenseQuestionPromptTemplate, ChatQuestionAnswerPromptTemplate } from '../../../prompt/templates/chat/index.js';
import { ModelAdapter } from '../../adapter.js';

describe('models/llms/meta/llama2', () => {
  describe('adapter', () => {
    const adapter = new ModelAdapter(LLAMA2_ADAPTER);

    test('should render qa prompt', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        ...adapter.prompt?.chat?.questionAnswer,
      });
      expect(await template.format({
        question: 'THE QUESTION',
        domain: 'DOMAIN',
        context: 'THE CORPUS',
      })).toMatchSnapshot();
    });

    test('should render condense prompt', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        ...adapter.prompt?.chat?.condenseQuestion,
      });
      expect(await template.format({
        question: 'THE QUESTION',
        chat_history: [
          new SystemMessage('A system message'),
          new HumanMessage('Hi'),
          new AIMessage('Hello!'),
        ],
      })).toMatchSnapshot();
    });
  });
});
