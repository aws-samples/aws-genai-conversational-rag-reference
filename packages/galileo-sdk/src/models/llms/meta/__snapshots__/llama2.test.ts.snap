// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/llms/meta/llama2 adapter should render condense prompt 1`] = `
"Given the following conversational dialog denoted by ''', and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.

Dialog: '''
System: A system message

Human: Hi

Assistant: Hello!
'''

Followup Question: THE QUESTION

Standalone Question: "
`;

exports[`models/llms/meta/llama2 adapter should render qa prompt 1`] = `
"You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer


Corpus: '''
THE CORPUS
'''

Question: THE QUESTION

Answer: "
`;
